// drop-down/modules/chart-theme/theme-default.js
  // 默认主题对象
  let theme = {
    dark: { // 主题 - 黑色
      title: '黑色',
      color: ['#dd6b66', '#759aa0', '#e69d87', '#8dc1a9', '#ea7e53', '#eedd78', '#73a373', '#73b9bc', '#7289ab',
        '#91ca8c', '#f49f42'
      ],
    },
    vintage: { // 主题 - 古老
      title: '古老',
      color: ['#d87c7c', '#919e8b', '#d7ab82', '#6e7074', '#61a0a8', '#efa18d', '#787464', '#cc7e63', '#724e58',
        '#4b565b'
      ]
    },
    infographic: { // 主题 - 信息图
      title: '信息图',
      color: ['#C1232B', '#27727B', '#FCCE10', '#E87C25', '#B5C334', '#FE8463', '#9BCA63', '#FAD860', '#F3A43B',
        '#60C0DD', '#D7504B', '#C6E579', '#F4E001', '#F0805A', '#26C0C0'
      ]
    },
    macarons: { // 主题 - 马卡龙
      title: '马卡龙',
      color: ['#2ec7c9', '#b6a2de', '#5ab1ef', '#ffb980', '#d87a80', '#8d98b3', '#e5cf0d', '#97b552', '#95706d',
        '#dc69aa',
        '#07a2a4', '#9a7fd1', '#588dd5', '#f5994e', '#c05050', '#59678c', '#c9ab00', '#7eb00a', '#6f5553',
        '#c14089'
      ]
    },
    roma: { // 主题 - 罗马
      title: '罗马',
      color: ['#E01F54', '#001852', '#f5e8c8', '#b8d2c7', '#c6b38e',
        '#a4d8c2', '#f3d999', '#d3758f', '#dcc392', '#2e4783',
        '#82b6e9', '#ff6347', '#a092f1', '#0a915d', '#eaf889',
        '#6699FF', '#ff6666', '#3cb371', '#d5b158', '#38b6b6'
      ]
    },
    shine: { // 主题 - 阳光
      title: '阳光',
      color: ['#c12e34', '#e6b600', '#0098d9', '#2b821d',
        '#005eaa', '#339ca8', '#cda819', '#32a487'
      ]
    },

  };
  export default theme;
